File: src/Lexer.cpp
#include "Common.h"
#include "Lexer.h"

std::map<Lexer::Token::Type, std::string> Lexer::TYPE_NAME_MAPPING = {
  {Token::Type::ILLEGAL, "Illegal"},
  {Token::Type::IDENTIFIER, "Identifier"},
  {Token::Type::KEYWORD, "Keyword"},
  {Token::Type::LITERAL, "Literal"},
  {Token::Type::MARKER, "Marker"},
  {Token::Type::OPERATOR, "Operator"},
};

std::map<std::string, Lexer::Token::Keyword> Lexer::KEYWORD_MAPPING = {
  {"val", Token::Keyword::CONSTANT},
  {"var", Token::Keyword::VARIABLE},
  {"fn", Token::Keyword::FUNCTION},
  {"return", Token::Keyword::RETURN},
  {"import", Token::Keyword::IMPORT},
  {"from", Token::Keyword::FROM},
  {"as", Token::Keyword::AS},
  {"struct", Token::Keyword::STRUCT},
  {"static", Token::Keyword::STATIC},
  {"pub", Token::Keyword::PUB},
  {"enum", Token::Keyword::ENUM},
  {"if", Token::Keyword::IF},
  {"else", Token::Keyword::ELSE},
  {"in", Token::Keyword::IN},
  {"match", Token::Keyword::MATCH},
  {"case", Token::Keyword::CASE},
  {"for", Token::Keyword::LOOP},
  {"continue", Token::Keyword::CONTINUE},
  {"break", Token::Keyword::BREAK},
};

std::map<char, Lexer::Token::Marker> Lexer::MARKER_MAPPING = {
  {'{', Token::Marker::BLOCK_BEGIN},
  {'}', Token::Marker::BLOCK_END},
  {'[', Token::Marker::BRACKET_BEGIN},
  {']', Token::Marker::BRACKET_END},
  {',', Token::Marker::COMMA},
  {'#', Token::Marker::COMMENT},
  {'(', Token::Marker::PARENTHESIS_BEGIN},
  {')', Token::Marker::PARENTHESIS_END},
  {'"', Token::Marker::STR_QUOTE},
};

std::map<std::string, Lexer::Token::Operator> Lexer::OPERATOR_MAPPING = {
  {"=", Token::Operator::ASSIGNMENT},
  {"+", Token::Operator::ADDITION},
  {"+=", Token::Operator::ADDITION_ASSIGNMENT},
  {"-", Token::Operator::SUBTRACTION},
  {"-=", Token::Operator::SUBTRACTION_ASSIGNMENT},
  {"*", Token::Operator::MULTIPLICATION},
  {"*=", Token::Operator::MULTIPLICATION_ASSIGNMENT},
  {"/", Token::Operator::DIVISION},
  {"/=", Token::Operator::DIVISION_ASSIGNMENT},
  {"%", Token::Operator::MODULUS},
  {"%=", Token::Operator::MODULUS_ASSIGNMENT},
  {"<", Token::Operator::LESS_THAN},
  {"<=", Token::Operator::LESS_THAN_EQUAL},
  {">", Token::Operator::GREATER_THAN},
  {">=", Token::Operator::GREATER_THAN_EQUAL},
  {"!=", Token::Operator::NOT_EQUAL},
  {"==", Token::Operator::EQUAL},
  {"and", Token::Operator::AND},
  {"or", Token::Operator::OR},
  {"not", Token::Operator::NOT},
  {":", Token::Operator::MEMBER_ACCESS},
  {"::", Token::Operator::STATIC_MEMBER_ACCESS},
};

Lexer::Token::Token(const std::string &value, const Type &type) {
  this->value = value;
  this->type = type;
}

Lexer::Token::Token(const std::string &value, const Type &type, const std::vector<std::string> &injections) {
  this->value = value;
  this->type = type;
  this->injections = injections;
}

Lexer::Token::Type Lexer::Token::get_type() const {
  return this->type;
}

std::string Lexer::Token::get_value() const {
  return this->value;
}

Lexer::Token::Keyword Lexer::Token::get_keyword() const {
  return KEYWORD_MAPPING.at(this->value);
}

bool Lexer::Token::is_given_marker(Marker marker) const {
  return type == Type::MARKER && MARKER_MAPPING.at(this->value[0]) == marker;
}

bool Lexer::Token::is_given_operator(Operator operation) const {
  return type == Type::OPERATOR && OPERATOR_MAPPING.at(this->value) == operation;
}

bool Lexer::Token::is_given_type(Type type_a, Type type_b) const {
  return type == type_a || type == type_b;
}

Lexer::Token Lexer::consume_buffer(std::string &buffer) {
  Token::Type type;

  if (is_boolean_literal(buffer)) {
    type = Token::Type::LITERAL;
  } else if (is_float_literal(buffer)) {
    type = Token::Type::LITERAL;
  } else if (is_integer_literal(buffer)) {
    type = Token::Type::LITERAL;
  } else if (is_keyword(buffer)) {
    type = Token::Type::KEYWORD;
  } else if (is_operator(buffer)) {
    type = Token::Type::OPERATOR;
  } else if (is_identifier(buffer)) {
    type = Token::Type::IDENTIFIER;
  } else {
    type = Token::Type::ILLEGAL;
  }

  Token token(buffer, type);
  buffer = "";
  return token;
}

std::string Lexer::consume_str_injection(const std::string &line, int &index) {
  std::string injection = "";

  for (int i = index + 1; i < line.size(); i++) {
    const char &character = line[i];

    if (is_identifier(character)) {
      injection += character;
      continue;
    }

    index = i - 1;
    return injection;
  }

  throw std::runtime_error("Unterminated String Injection");
}

Lexer::Token Lexer::build_str_literal(const std::string &line, int &index) {
  std::string buffer = "";
  std::vector<std::string> injections;
  
  for (int i = index + 1; i < line.size(); i++) {
    const char &character = line[i];

    if (character == '"') {
      index = i;
      return Token(buffer, Token::Type::LITERAL, injections);
    }

    if (is_str_injection(line, i)) {
      std::string injection = consume_str_injection(line, i);
      injections.push_back(injection);
      buffer += "#" + injection;
      continue;
    }

    buffer += character;
  }

  throw std::runtime_error("Unterminated String Literal");
}

Lexer::Token Lexer::build_operator(const std::string &buffer, int &index) {
  std::string single_character_operator = std::string(1, buffer[index]);

  for (int i = index + 1; i < buffer.size(); i++) {
    const char &character = buffer[i];

    if (isspace(character)) continue;

    std::string character_str = std::string(1, character);

    if (is_operator(character_str) or character_str == "!") {
      std::string dual_character_operator = single_character_operator + character_str;

      index = i;
      return Token(
        dual_character_operator,
        is_operator(dual_character_operator) ? Token::Type::OPERATOR : Token::Type::ILLEGAL
      );
    } else {
      break;
    }
  }

  return Token(
    single_character_operator, 
    is_operator(single_character_operator) ? Token::Type::OPERATOR : Token::Type::ILLEGAL
  );
}

void Lexer::Token::print() const {
  println("Token {");
  println("  type: " + TYPE_NAME_MAPPING.at(this->type));
  println("  value: " + this->value);
  if (not this->injections.empty()) {
    println("  injections: [");
    for (const std::string &injection : this->injections) {
      println("    " + injection);
    }
    println("  ]");
  }
  println("}");
}

Lexer::Stream::Stream(const std::vector<Token> &collection) {
  this->index = 0;
  this->collection = std::move(collection);
}

const Lexer::Token& Lexer::Stream::current() {
  return this->collection[this->index];
}

const Lexer::Token& Lexer::Stream::consume() {
  return this->collection[this->index++];
}

void Lexer::Stream::next() {
  this->index++;
}

bool Lexer::Stream::has_next() const {
  return this->index < this->collection.size();
}

bool Lexer::Stream::is_next(const std::function<bool(const Token &)> &predicate) const {  
  return this->index + 1 < this->collection.size() and predicate(this->collection[this->index + 1]);
}

bool Lexer::is_boolean_literal(const std::string &buffer) {
  return buffer == "true" or buffer == "false";
}

bool Lexer::is_float_literal(const std::string &buffer) {
  bool has_decimal = false;

  for (int i = 0; i < buffer.size(); i++) {
    const char &character = buffer[i];

    if (isdigit(character)) continue;
    if (character == '.') {
      if (has_decimal) throw std::runtime_error("Invalid Float Literal");
      has_decimal = true;
    } 
  }

  return has_decimal;
}

bool Lexer::is_integer_literal(const std::string &buffer) {
  return std::all_of(buffer.begin(), buffer.end(), [](const char &character) {
    return isdigit(character);
  });
}
bool Lexer::is_identifier(const char &character) {
  return isalnum(character) or character == '_' or character == '$';
}

bool Lexer::is_identifier(const std::string &buffer) {
  return not isdigit(buffer[0]) and std::all_of(buffer.begin(), buffer.end(), [](const char &character) {
    return is_identifier(character);
  });
}

bool Lexer::is_keyword(const std::string &buffer) {
  return KEYWORD_MAPPING.find(buffer) != KEYWORD_MAPPING.end();
}

bool Lexer::is_marker(const char &character) {
  return MARKER_MAPPING.find(character) != MARKER_MAPPING.end();
}

bool Lexer::is_next_char(const std::string &line, int index, std::function<bool(const char &)> predicate) {
  if (index + 1 > line.size()) return false;
  return predicate(line[index + 1]);
}

bool Lexer::is_operator(const std::string &buffer) {
  return OPERATOR_MAPPING.find(buffer) != OPERATOR_MAPPING.end();
}

bool Lexer::is_str_injection(const std::string &line, size_t index) {
  return line[index] == '#' and is_next_char(line, index, [](const char &character) {
    return not isdigit(character) and isalpha(character);
  });
}

Lexer::Token::Marker Lexer::get_marker(const char &character) {
  return MARKER_MAPPING.at(character);
}

std::vector<Lexer::Token> Lexer::lex(const std::string &line) {
  std::vector<Token> collection;
  std::string final_line = line + " ";
  std::string buffer = "";

  for (int i = 0; i < final_line.size(); i++) {
    const char &character = final_line[i];

    if (is_whitespace(character)) {
      if (is_whitespace(buffer)) continue;
      collection.push_back(consume_buffer(buffer));
      continue;
    }

    std::string character_str = std::string(1, character);
    if (is_operator(character_str) or character_str == "!") {
      if (not is_whitespace(buffer)) {
        collection.push_back(consume_buffer(buffer));
      }

      collection.push_back(
        build_operator(final_line, i)
      );
      continue;
    } 
    
    if (is_marker(character)) {
      if (not is_whitespace(buffer)) {
        collection.push_back(consume_buffer(buffer));
      }

      Token::Marker marker = get_marker(character);

      switch (marker) {
        case Token::Marker::STR_QUOTE:
          collection.push_back(build_str_literal(final_line, i));
          break;
        case Token::Marker::COMMENT:
          return collection;
        default:
          collection.push_back(Token(std::string(1, character), Token::Type::MARKER));
      }

      continue;
    }

    buffer += character;
  }

  return collection;
}

Lexer::Stream Lexer::lex_file(const std::string &filename) {
  std::vector<Token> collection;

  each_line(filename, [&collection](const std::string &line) {
    std::vector<Token> tokens = lex(line);
    collection.insert(collection.end(), tokens.begin(), tokens.end());
  });

  return Stream(collection);
}

File: pino-csharp/Lexer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace Pino;

public class Lexer {
  private static readonly Dictionary<string, KeywordType> Keywords = new() {
        { "as", KeywordType.As },
        { "break", KeywordType.Break },
        { "val", KeywordType.Constant },
        { "continue", KeywordType.Continue },
        { "else", KeywordType.Else },
        { "enum", KeywordType.Enum },
        { "from", KeywordType.From },
        { "fn", KeywordType.Function },
        { "if", KeywordType.If },
        { "import", KeywordType.Import },
        { "in", KeywordType.In },
        { "interface", KeywordType.Interface },
        { "is", KeywordType.Is },
        { "for", KeywordType.Loop },
        { "match", KeywordType.Match },
        { "module", KeywordType.Module },
        { "pub", KeywordType.Pub },
        { "return", KeywordType.Return },
        { "static", KeywordType.Static },
        { "struct", KeywordType.Struct },
        { "then", KeywordType.Then },
        { "var", KeywordType.Variable },
        { "when", KeywordType.When }
    };

  private static readonly Dictionary<char, MarkerType> Markers = new() {
        { '@', MarkerType.At },
        { '{', MarkerType.BlockBegin },
        { '}', MarkerType.BlockEnd },
        { '[', MarkerType.BracketBegin },
        { ']', MarkerType.BracketEnd },
        { ',', MarkerType.Comma },
        { '#', MarkerType.Comment },
        { '(', MarkerType.ParenthesisBegin },
        { ')', MarkerType.ParenthesisEnd },
        { '"', MarkerType.StrQuote }
    };

  private static readonly Dictionary<string, OperatorType> Operators = new() {
        { "=", OperatorType.Assignment },
        { "+", OperatorType.Addition },
        { "+=", OperatorType.AdditionAssignment },
        { "-", OperatorType.Subtraction },
        { "-=", OperatorType.SubtractionAssignment },
        { "*", OperatorType.Multiplication },
        { "*=", OperatorType.MultiplicationAssignment },
        { "/", OperatorType.Division },
        { "/=", OperatorType.DivisionAssignment },
        { "%", OperatorType.Modulus },
        { "%=", OperatorType.ModulusAssignment },
        { "<", OperatorType.LessThan },
        { "<=", OperatorType.LessThanEqual },
        { ">", OperatorType.GreaterThan },
        { ">=", OperatorType.GreaterThanEqual },
        { "==", OperatorType.Equal },
        { "!=", OperatorType.NotEqual },
        { "and", OperatorType.And },
        { "or", OperatorType.Or },
        { "not", OperatorType.Not },
        { ":", OperatorType.MemberAccess },
        { "=>", OperatorType.Arrow },
        { "::", OperatorType.StaticMemberAccess }
    };

  private static readonly Regex FloatRegex = new(@"^-?[0-9]{1,9}(_[0-9]{3})*\.[0-9]{1,9}(_[0-9]{3})*$");
  private static readonly Regex IntRegex = new(@"^-?[0-9]{1,18}(_[0-9]{3})*$");
  private static readonly Regex IdentifierRegex = new(@"^[a-zA-Z_$][a-zA-Z0-9_$]*$");

  public static List<Token> LexLine(string line, int lineNum = 1) {
    var tokens = new List<Token>();
    var index = 0;

    while (index < line.Length) {
      var c = line[index];

      // 1. Whitespace
      if (char.IsWhiteSpace(c)) {
        index++;
        continue;
      }

      // 2. Comments
      if (c == '#') {
        // Comment spans to the end of the line
        break;
      }

      // 3. String Quote (String Literals with Potential Injections)
      if (c == '"') {
        LexStringLiteral(line, ref index, tokens, lineNum);
        continue;
      }

      // 3b. Rune Quote
      if (c == '\'') {
        LexRuneLiteral(line, ref index, tokens, lineNum);
        continue;
      }

      // 4. Markers (except comment, which is handled above)
      if (Markers.TryGetValue(c, out var markerType) && markerType != MarkerType.Comment && markerType != MarkerType.StrQuote) {
        var token = new Token(TokenType.Marker, c.ToString(), Marker: markerType) { Line = lineNum, Column = index + 1 };
        tokens.Add(token);
        index++;
        continue;
      }

      // 5. Multi-character Operators (like ==, !=, +=, -=, ::, <=, >=)
      if (index + 1 < line.Length) {
        var dualOp = line.Substring(index, 2);
        if (Operators.TryGetValue(dualOp, out var dualOpType) && !char.IsLetter(dualOp[0])) {
          var token = new Token(TokenType.Operator, dualOp, Operator: dualOpType) { Line = lineNum, Column = index + 1 };
          tokens.Add(token);
          index += 2;
          continue;
        }
      }

      // Single-character Operators
      if (Operators.TryGetValue(c.ToString(), out var opType) && c != 'a' && c != 'o' && c != 'n') { // avoid partial "and", "or", "not"
        var token = new Token(TokenType.Operator, c.ToString(), Operator: opType) { Line = lineNum, Column = index + 1 };
        tokens.Add(token);
        index++;
        continue;
      }

      // 6. Alphanumeric / Identifier / Number / Keyword buffer
      var bufferStart = index;
      while (index < line.Length) {
        var nextChar = line[index];
        if (char.IsWhiteSpace(nextChar) || Markers.ContainsKey(nextChar) || (Operators.ContainsKey(nextChar.ToString()) && nextChar != '_')) {
          break;
        }
        index++;
      }

      var buffer = line.Substring(bufferStart, index - bufferStart);
      if (!string.IsNullOrEmpty(buffer)) {
        var token = GetTokenFromBuffer(buffer);
        token.Line = lineNum;
        token.Column = bufferStart + 1;
        tokens.Add(token);
      }
    }

    return tokens;
  }

  public static List<Token> LexFile(string filePath) {
    var tokens = new List<Token>();
    int lineNum = 1;
    var lines = File.ReadAllLines(filePath);
    PinoException.RegisterFileLines(filePath, lines);
    foreach (var line in lines) {
      tokens.AddRange(LexLine(line, lineNum));
      lineNum++;
    }
    return tokens;
  }

  private static void LexStringLiteral(string line, ref int index, List<Token> tokens, int lineNum) {
    var buffer = new StringBuilder();
    bool hasAddedInitialString = false;
    int colStart = index + 1;
    index++; // Skip opening quote

    while (index < line.Length) {
      var c = line[index];

      // Check for closing quote, ignoring escaped quotes \"
      if (c == '"' && (index == 0 || line[index - 1] != '\\')) {
        index++; // Skip closing quote
        if (!hasAddedInitialString) {
          var token = new Token(TokenType.Literal, buffer.ToString(), Literal: LiteralType.String) { Line = lineNum, Column = colStart };
          tokens.Add(token);
        } else if (buffer.Length > 0) {
          var opToken = new Token(TokenType.Operator, "+", Operator: OperatorType.Addition) { Line = lineNum, Column = index };
          tokens.Add(opToken);
          var token = new Token(TokenType.Literal, buffer.ToString(), Literal: LiteralType.String) { Line = lineNum, Column = index };
          tokens.Add(token);
        }
        return;
      }

      // Check for string injection $(expr) or $varName, ignoring escaped \$
      if (c == '$' && (index == 0 || line[index - 1] != '\\') && index + 1 < line.Length) {
        if (line[index + 1] == '(') {
          // Complex interpolation $(expr)
          if (!hasAddedInitialString) {
            var token = new Token(TokenType.Literal, buffer.ToString(), Literal: LiteralType.String) { Line = lineNum, Column = colStart };
            tokens.Add(token);
            hasAddedInitialString = true;
          } else if (buffer.Length > 0) {
            var opToken = new Token(TokenType.Operator, "+", Operator: OperatorType.Addition) { Line = lineNum, Column = index + 1 };
            tokens.Add(opToken);
            var token = new Token(TokenType.Literal, buffer.ToString(), Literal: LiteralType.String) { Line = lineNum, Column = index + 1 };
            tokens.Add(token);
          }
          buffer.Clear();

          var outerOp = new Token(TokenType.Operator, "+", Operator: OperatorType.Addition) { Line = lineNum, Column = index + 1 };
          tokens.Add(outerOp);

          int parenCol = index + 2;
          index += 2; // skip "$("

          int depth = 1;
          int startExpr = index;
          while (index < line.Length && depth > 0) {
            char ec = line[index];
            if (ec == '"') {
              index++; // skip opening quote
              while (index < line.Length) {
                if (line[index] == '"' && line[index - 1] != '\\') {
                  break;
                }
                index++;
              }
            } else if (ec == '(') {
              depth++;
            } else if (ec == ')') {
              depth--;
            }
            index++;
          }

          string exprStr = line.Substring(startExpr, index - 1 - startExpr);
          var exprTokens = LexLine(exprStr, lineNum);
          foreach (var et in exprTokens) {
            et.Column += startExpr;
          }
          var pBegin = new Token(TokenType.Marker, "(", Marker: MarkerType.ParenthesisBegin) { Line = lineNum, Column = parenCol };
          tokens.Add(pBegin);
          tokens.AddRange(exprTokens);
          var pEnd = new Token(TokenType.Marker, ")", Marker: MarkerType.ParenthesisEnd) { Line = lineNum, Column = index };
          tokens.Add(pEnd);
          continue;
        } else if (IsIdentifierStart(line[index + 1]) && !char.IsDigit(line[index + 1])) {
          // Simple interpolation $varName
          if (!hasAddedInitialString) {
            var token = new Token(TokenType.Literal, buffer.ToString(), Literal: LiteralType.String) { Line = lineNum, Column = colStart };
            tokens.Add(token);
            hasAddedInitialString = true;
          } else if (buffer.Length > 0) {
            var opToken = new Token(TokenType.Operator, "+", Operator: OperatorType.Addition) { Line = lineNum, Column = index + 1 };
            tokens.Add(opToken);
            var token = new Token(TokenType.Literal, buffer.ToString(), Literal: LiteralType.String) { Line = lineNum, Column = index + 1 };
            tokens.Add(token);
          }
          buffer.Clear();

          var outerOp = new Token(TokenType.Operator, "+", Operator: OperatorType.Addition) { Line = lineNum, Column = index + 1 };
          tokens.Add(outerOp);
          index++; // skip '$'
          int startIdent = index;
          while (index < line.Length && IsIdentifierChar(line[index])) {
            index++;
          }
          string identStr = line.Substring(startIdent, index - startIdent);
          var identToken = new Token(TokenType.Identifier, identStr) { Line = lineNum, Column = startIdent + 1 };
          tokens.Add(identToken);
          continue;
        }
      }

      buffer.Append(c);
      index++;
    }

    throw new Exception($"Unterminated string literal at line index {index}");
  }

  private static Token GetTokenFromBuffer(string buffer) {
    if (Keywords.TryGetValue(buffer, out var keywordType)) {
      return new Token(TokenType.Keyword, buffer, Keyword: keywordType);
    }

    if (Operators.TryGetValue(buffer, out var opType)) {
      return new Token(TokenType.Operator, buffer, Operator: opType);
    }

    if (buffer == "true" || buffer == "false") {
      return new Token(TokenType.Literal, buffer, Literal: LiteralType.Boolean);
    }

    if (IntRegex.IsMatch(buffer)) {
      return new Token(TokenType.Literal, buffer, Literal: LiteralType.Integer);
    }

    if (FloatRegex.IsMatch(buffer)) {
      return new Token(TokenType.Literal, buffer, Literal: LiteralType.Float);
    }

    if (IdentifierRegex.IsMatch(buffer)) {
      return new Token(TokenType.Identifier, buffer);
    }

    return new Token(TokenType.Illegal, buffer);
  }

  private static bool IsIdentifierStart(char c) {
    return char.IsLetter(c) || c == '_' || c == '$';
  }

  private static bool IsIdentifierChar(char c) {
    return char.IsLetterOrDigit(c) || c == '_' || c == '$';
  }

  private static void LexRuneLiteral(string line, ref int index, List<Token> tokens, int lineNum) {
    int start = index;
    int colStart = index + 1;
    index++; // Skip opening single quote '
    if (index >= line.Length) {
      throw new Exception("Lexer Error: Unterminated rune literal");
    }

    int codePoint = 0;
    if (line[index] == '\\') {
      index++; // Skip backslash
      if (index >= line.Length) {
        throw new Exception("Lexer Error: Unterminated escape sequence in rune literal");
      }
      char escape = line[index];
      codePoint = escape switch {
        'n' => '\n',
        't' => '\t',
        'r' => '\r',
        '\'' => '\'',
        '\\' => '\\',
        _ => throw new Exception($"Lexer Error: Invalid escape character '{escape}' in rune literal")
      };
      index++;
    } else {
      if (char.IsHighSurrogate(line[index]) && index + 1 < line.Length && char.IsLowSurrogate(line[index + 1])) {
        codePoint = char.ConvertToUtf32(line[index], line[index + 1]);
        index += 2;
      } else {
        codePoint = line[index];
        index++;
      }
    }

    if (index >= line.Length || line[index] != '\'') {
      throw new Exception($"Lexer Error: Rune literal must end with a single quote: {line.Substring(start, Math.Min(line.Length - start, 5))}");
    }
    index++; // Skip closing single quote
    var token = new Token(TokenType.Literal, codePoint.ToString(), Literal: LiteralType.Rune) { Line = lineNum, Column = colStart };
    tokens.Add(token);
  }
}
